### PR TITLE
Include licenses  and recipricol source-files in OTEL's docker distro

### LINKED ISSUE #2458: Add 3rd-party LICENSE files to Docker Image distribution
**Is your feature request related to a problem? Please describe.**
Many OSS licenses require re-distribution with said license.   E.g. the MIT licenses explicitly states:

`The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.`

**Describe the solution you'd like**
We create a `licenses/` directory in the relevant distributed DOCKER files that contains all 3p licenses.

Ideally the following would work:

```
docker create --name="tmp_collector" otel/opentelemetry-collector
docker export tmp_collector | tar -x licenses
docker rm tmp_collector
cat licenses/*
```

(Note: I can submit a PR).

**Describe alternatives you've considered**
I didn't consider other alternatives to distributing the license. I guess you could encode it in the binary and dump it in logs at startup?

### PR DESCRIPTION
**Pull request by jsuereth**
Fixes #2458

- Adds `go-licenses` tool to `make install-tools`
- Adds a new `otelcol-licenses` make target which dumps all license files from dependencies used to build `otelcol`
  -  This includes the original license in a directory associated with the go dependency
  - For "reciprocal" licenses (weak copyleft, like MPL) that require access to source, this copies the source used *at build time* into the docker image to comply with the license.
- Updated docker image to include this "third-party" directory, complying with distribution requirements on 3P licenses.

### CODE REVIEW

**Comment by tigrannajaryan** [comment]
>For "reciprocal" licenses (weak copyleft, like MPL) that require access to source, this copies the source used at build time into the docker image to comply with the license.

Do we have licenses like this? Can you show the output of the tool?

**Comment by jsuereth** [comment]
@tigrannajaryan 

> Do we have licenses like this? Can you show the output of the tool?
[MPL](https://opensource.org/licenses/MPL-2.0) is a "weak copyleft" license, and yes, I think most hashicorp libraries are MPL 2.0.

The go-licenses tool takes the "strictest" definition of MPL v2.0 paragraph 3.2 and just bundles the source right in the executable.    I'm not aware of another tool that copy-pastes the LICENSE notices in a clear way for dependencies (but am totally happy to use another tool if you know of one).

Here's a (truncated) example where I highlight what's done with MPL dependencies:

```
$ ls -R third-party/*
third-party/go.mod

third-party/cloud.google.com:
go

third-party/cloud.google.com/go:
compute

third-party/cloud.google.com/go/compute:
metadata

third-party/cloud.google.com/go/compute/metadata:
LICENSE

third-party/contrib.go.opencensus.io:
exporter

third-party/contrib.go.opencensus.io/exporter:
prometheus

third-party/contrib.go.opencensus.io/exporter/prometheus:
LICENSE

...

third-party/github.com/hashicorp/consul/api:
acl.go         catalog.go                           config_entry.go                  connect.go                 debug_test.go            go.sum          lock.go                operator_autopilot_test.go  operator_raft_test.go   README.md          snapshot_test.go
acl_test.go    catalog_test.go                      config_entry_intentions.go       connect_intention.go       discovery_chain.go       health.go       lock_test.go           operator.go                 operator_segment.go     semaphore.go       status.go
agent.go       config_entry_discoverychain.go       config_entry_intentions_test.go  connect_intention_test.go  discovery_chain_test.go  health_test.go  namespace.go           operator_keyring.go         oss_test.go             semaphore_test.go  status_test.go
agent_test.go  config_entry_discoverychain_test.go  config_entry_test.go             coordinate.go              event.go                 kv.go           namespace_test.go      operator_keyring_test.go    prepared_query.go       session.go         txn.go
api.go         config_entry_gateways.go             connect_ca.go                    coordinate_test.go         event_test.go            kv_test.go      operator_area.go       operator_license.go         prepared_query_test.go  session_test.go    txn_test.go
api_test.go    config_entry_gateways_test.go        connect_ca_test.go               debug.go                   go.mod                   LICENSE         operator_autopilot.go  operator_raft.go            raw.go                  snapshot.go        watch

third-party/github.com/hashicorp/consul/api/watch:
funcs.go  funcs_test.go  plan.go  plan_test.go  watch.go  watch_test.go


...
```

**Comment by tigrannajaryan** [comment]
We need a clarification on MPL, commented here: https://github.com/open-telemetry/community/issues/649#issuecomment-792892303

**Comment by jsuereth** [comment]
@tigrannajaryan Should I be working to fix this up, or are we still waiting on the MPL question?

**Comment by tigrannajaryan** [comment]
> @tigrannajaryan Should I be working to fix this up, or are we still waiting on the MPL question?

I do not expect that MPL may be prohibited, so I believe you can work on this in parallel. We do need to clarify the license question though (if it turns out that MPL is prohibited we will need to find replacements for dependencies).

**Comment by tigrannajaryan** [comment]
I opened a separate issue to clarify MPL dependency usage: https://github.com/open-telemetry/community/issues/688

**Comment by punya** [comment]
@jsuereth should we pick this back up?

File: Makefile [hunk 1]
```diff
@@ -154,13 +155,21 @@ otelcol:
 	go generate ./...
 	$(MAKE) build-binary-internal
 
+.PHONY: otelcol-licenses
+otelcol-licenses: cmd/otelcol/third-party
+
+cmd/otelcol/third-party:
+	@echo creating third_party directory with licenses + reciprical source
+	cd ./cmd/otelcol && go-licenses save . --save_path=third-party
+	chmod +w $(find ./cmd/otelcol/third_party -type d)
```
Comment (jrcamp) [inline Makefile:164]: Had some issues running this due to makefile target and paths not matching up (third_party/third-party). This worked for me:

```suggestion
.PHONY: otelcol-licenses
otelcol-licenses:
	@echo creating third_party directory with licenses + reciprical source
	go-licenses save ./cmd/otelcol --save_path=./cmd/otelcol/third-party --force
	chmod -R u+w ./cmd/otelcol/third-party
```

Was the `go.mod` to get rid of messages like
> main module (go.opentelemetry.io/collector) does not contain package go.opentelemetry.io/collector/cmd/otelcol/third-party/github.com/hashicorp/consul/api

I get those as well but they seem harmless. Would be nice not to have them but 🤷

**Comment by barney-s** [comment]
@jsuereth - Would you have time to resolve the comments. Else i can pick this up and take it closure.

File: Makefile (hunk 1 above)
Comment (jsuereth) [reply Makefile]: Ok, so I fixed the typo (sorry bout that), and I've been upgrading.  Seems we still get issues here so I'm toying around with some solutions.

**Comment by jsuereth** [comment]
@tigrannajaryan This is ready for another look. The license cleanup looks good to me, do you want a dump of what is included now?

**Comment by tigrannajaryan** [comment]
> @tigrannajaryan This is ready for another look. The license cleanup looks good to me, do you want a dump of what is included now?

Yes, let's see the dump.

For MPL licenses https://github.com/cncf/foundation/issues/138 we need to ask for exceptions for each individual dependency. They said:

>As a followup, the GB decided to not grant a blanket exception to hashi
libs but will approve things on an individual basis.

**Comment by jsuereth** [comment]
Full dump of license files (not seeing any that require source this time):

```
third-party/*
third-party/contrib.go.opencensus.io:
exporter

third-party/contrib.go.opencensus.io/exporter:
prometheus

third-party/contrib.go.opencensus.io/exporter/prometheus:
LICENSE

third-party/github.com:
beorn7  cenkalti  cespare  felixge  fsnotify  gogo  go-kit  golang  go-logfmt  google  gorilla  knadh  magiconair  matttproud  mitchellh  prometheus  rs  shirou  spf13  tklauser

third-party/github.com/beorn7:
perks

third-party/github.com/beorn7/perks:
quantile

third-party/github.com/beorn7/perks/quantile:
LICENSE

third-party/github.com/cenkalti:
backoff

third-party/github.com/cenkalti/backoff:
v4

third-party/github.com/cenkalti/backoff/v4:
LICENSE

third-party/github.com/cespare:
xxhash

third-party/github.com/cespare/xxhash:
v2

third-party/github.com/cespare/xxhash/v2:
LICENSE.txt

third-party/github.com/felixge:
httpsnoop

third-party/github.com/felixge/httpsnoop:
LICENSE.txt

third-party/github.com/fsnotify:
fsnotify

third-party/github.com/fsnotify/fsnotify:
LICENSE

third-party/github.com/gogo:
protobuf

third-party/github.com/gogo/protobuf:
LICENSE

third-party/github.com/go-kit:
log

third-party/github.com/go-kit/log:
LICENSE

third-party/github.com/golang:
groupcache  protobuf

third-party/github.com/golang/groupcache:
lru

third-party/github.com/golang/groupcache/lru:
LICENSE

third-party/github.com/golang/protobuf:
LICENSE

third-party/github.com/go-logfmt:
logfmt

third-party/github.com/go-logfmt/logfmt:
LICENSE

third-party/github.com/google:
uuid

third-party/github.com/google/uuid:
LICENSE

third-party/github.com/gorilla:
mux

third-party/github.com/gorilla/mux:
LICENSE

third-party/github.com/knadh:
koanf

third-party/github.com/knadh/koanf:
LICENSE

third-party/github.com/magiconair:
properties

third-party/github.com/magiconair/properties:
LICENSE.md

third-party/github.com/matttproud:
golang_protobuf_extensions

third-party/github.com/matttproud/golang_protobuf_extensions:
pbutil

third-party/github.com/matttproud/golang_protobuf_extensions/pbutil:
LICENSE  NOTICE

third-party/github.com/mitchellh:
copystructure  mapstructure  reflectwalk

third-party/github.com/mitchellh/copystructure:
LICENSE

third-party/github.com/mitchellh/mapstructure:
LICENSE

third-party/github.com/mitchellh/reflectwalk:
LICENSE

third-party/github.com/prometheus:
client_golang  client_model  common  procfs  statsd_exporter

third-party/github.com/prometheus/client_golang:
prometheus

third-party/github.com/prometheus/client_golang/prometheus:
LICENSE  NOTICE

third-party/github.com/prometheus/client_model:
go

third-party/github.com/prometheus/client_model/go:
LICENSE  NOTICE

third-party/github.com/prometheus/common:
internal  LICENSE  NOTICE

third-party/github.com/prometheus/common/internal:
bitbucket.org

third-party/github.com/prometheus/common/internal/bitbucket.org:
ww

third-party/github.com/prometheus/common/internal/bitbucket.org/ww:
goautoneg

third-party/github.com/prometheus/common/internal/bitbucket.org/ww/goautoneg:
README.txt

third-party/github.com/prometheus/procfs:
LICENSE  NOTICE

third-party/github.com/prometheus/statsd_exporter:
pkg

third-party/github.com/prometheus/statsd_exporter/pkg:
mapper

third-party/github.com/prometheus/statsd_exporter/pkg/mapper:
LICENSE  NOTICE

third-party/github.com/rs:
cors

third-party/github.com/rs/cors:
LICENSE

third-party/github.com/shirou:
gopsutil

third-party/github.com/shirou/gopsutil:
LICENSE

third-party/github.com/spf13:
cast  cobra  pflag

third-party/github.com/spf13/cast:
LICENSE

third-party/github.com/spf13/cobra:
LICENSE.txt

third-party/github.com/spf13/pflag:
LICENSE

third-party/github.com/tklauser:
go-sysconf  numcpus

third-party/github.com/tklauser/go-sysconf:
LICENSE

third-party/github.com/tklauser/numcpus:
LICENSE

third-party/golang.org:
x

third-party/golang.org/x:
net  sys  text

third-party/golang.org/x/net:
LICENSE

third-party/golang.org/x/sys:
LICENSE

third-party/golang.org/x/text:
LICENSE

third-party/google.golang.org:
genproto  grpc  protobuf

third-party/google.golang.org/genproto:
googleapis

third-party/google.golang.org/genproto/googleapis:
rpc

third-party/google.golang.org/genproto/googleapis/rpc:
LICENSE

third-party/google.golang.org/grpc:
LICENSE

third-party/google.golang.org/protobuf:
LICENSE

third-party/go.opencensus.io:
LICENSE

third-party/go.opentelemetry.io:
collector  contrib  otel

third-party/go.opentelemetry.io/collector:
LICENSE

third-party/go.opentelemetry.io/contrib:
instrumentation  LICENSE  zpages

third-party/go.opentelemetry.io/contrib/instrumentation:
google.golang.org  net

third-party/go.opentelemetry.io/contrib/instrumentation/google.golang.org:
grpc

third-party/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc:
otelgrpc

third-party/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc:
LICENSE

third-party/go.opentelemetry.io/contrib/instrumentation/net:
http

third-party/go.opentelemetry.io/contrib/instrumentation/net/http:
otelhttp

third-party/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp:
LICENSE

third-party/go.opentelemetry.io/contrib/zpages:
LICENSE

third-party/go.opentelemetry.io/otel:
internal  LICENSE  metric  sdk  trace

third-party/go.opentelemetry.io/otel/internal:
metric

third-party/go.opentelemetry.io/otel/internal/metric:
global

third-party/go.opentelemetry.io/otel/internal/metric/global:
LICENSE

third-party/go.opentelemetry.io/otel/metric:
LICENSE

third-party/go.opentelemetry.io/otel/sdk:
LICENSE

third-party/go.opentelemetry.io/otel/trace:
LICENSE

third-party/gopkg.in:
yaml.v2

third-party/gopkg.in/yaml.v2:
LICENSE  NOTICE

third-party/go.uber.org:
atomic  multierr  zap

third-party/go.uber.org/atomic:
LICENSE.txt

third-party/go.uber.org/multierr:
LICENSE.txt

third-party/go.uber.org/zap:
LICENSE.txt

third-party/vendor:
golang.org

third-party/vendor/golang.org:
x

third-party/vendor/golang.org/x:
crypto  net  sys  text

third-party/vendor/golang.org/x/crypto:
LICENSE

third-party/vendor/golang.org/x/net:
LICENSE

third-party/vendor/golang.org/x/sys:
cpu

third-party/vendor/golang.org/x/sys/cpu:
LICENSE

third-party/vendor/golang.org/x/text:
LICENSE
```

I can dump the licenses themselves, but stretching the limit of github :)

**Comment by jsuereth** [comment]
@tigrannajaryan let me know if you need any more insight into what this is doing, or restructuring.

File: cmd/otelcol/Dockerfile [hunk 2]
```diff
@@ -6,6 +6,10 @@ COPY otelcol /
 # Note that this shouldn't be necessary, but in some cases the file seems to be
 # copied with the execute bit lost (see #1317)
 RUN chmod 755 /otelcol
```
Comment (bogdandrutu) [inline cmd/otelcol/Dockerfile:8]: 750?

**Review by bogdandrutu — APPROVED** [review]

File: cmd/otelcol/Dockerfile (hunk 2 above)
Comment (jsuereth) [reply cmd/otelcol/Dockerfile:8]: This is what it was.  IF you think the chmod of the default docker file needs to change, then perhaps that deserves a separate bug/discussion.

**Comment by bogdandrutu** [comment]
@jsuereth we no-longer build the collector with "make" directly but we use the builder, which I think will benefit everyone to have this in the builder.

@jpkrohling can we add this to the builder to add a all "licenses" file in the docker image? Would be great I think.

**Comment by jpkrohling** [comment]
I believe this PR here should be closed in favor of doing the same on the releases repository.

**Pull request closed without merging**